Add DND (Do Not Disturb) Integration

MachineThing opened the pull request:
This pull request adds these following changes to the app:

- Do not disturb integration, when the user starts tracking their sleep the app turns DND mode automatically and turns it to the last state the user left DND on when they stop tracking sleep
- Added DND setting to the preferences, it also asks the user for the DND permission which is required for DND integration to work.

This might not be the perfect way to do this but it does seem to work well for me at least (on emulator and real phone), and this pull request also solves issue #419

vmiklos commented:
Hi, this looks quite promising. I like how it default to false, so existing happy users are left alone.

Could you please fix up the ktlint problems? See the CI log.

Apart from that, if you could write a paragraph in guide/ about the feature, that would be also useful.

Thanks!

MachineThing commented:
I fixed the KtLint issues and added a small paragraph on the function.

> I like how it defaults to false, so existing happy users are left alone.

By the way if your curious it defaults to false because the app needs to ask the user for permissions to make it so the app can turn DND on and off, so it prompts the user to give it permissions when toggled on and the app notices it doesn't have permissions. :smile:

vmiklos commented on app/src/main/java/hu/vmiklos/plees_tracker/MainActivity.kt:
```diff
@@ -313,7 +315,20 @@ class MainActivity : AppCompatActivity(), View.OnClickListener {
         if (DataModel.start != null && DataModel.stop != null) {
+            // When user stops tracking sleep
+            if (dndManager.isNotificationPolicyAccessGranted and dndEnabled) {
+                // Restore Do Not Disturb status when user started tracking
```
This is a bit misleading,  you restore it when you stop, you just restore it to the value that was set when you started, right?

vmiklos commented on app/src/main/java/hu/vmiklos/plees_tracker/MainActivity.kt:
```diff
@@ -313,7 +315,20 @@ class MainActivity : AppCompatActivity(), View.OnClickListener {
+                val filterAll = NotificationManager.INTERRUPTION_FILTER_ALL
+                dndManager.setInterruptionFilter(preferences.getInt("current_dnd", filterAll))
+            } else {
+                Log.w(TAG, "Failed to disable DND, permissions not enabled")
```
Would be ideal to only warn when this is enabled but the access is not granted.

MachineThing replied:
True, line 324 used to be `if (dndManager.isNotificationPolicyAccessGranted)` before I added the ability to enable/disable the feature in the settings which is why it was like that.

vmiklos commented on app/src/main/java/hu/vmiklos/plees_tracker/MainActivity.kt:
```diff
@@ -325,7 +340,17 @@ class MainActivity : AppCompatActivity(), View.OnClickListener {
+            if (dndManager.isNotificationPolicyAccessGranted and dndEnabled) {
+                dndManager.setInterruptionFilter(NotificationManager.INTERRUPTION_FILTER_PRIORITY)
+            } else {
+                Log.w(TAG, "Failed to enable DND, permissions not enabled")
```
Same here, only warn when the setting is on, but we have no access.

MachineThing replied:
Ditto

vmiklos commented on app/src/main/java/hu/vmiklos/plees_tracker/Preferences.kt:
```diff
@@ -20,6 +30,76 @@ class Preferences : PreferenceFragmentCompat() {
+                                    editor.putBoolean("enable_dnd", false)
+                                    editor.apply()
+                                } else {
+                                    Log.wtf(TAG, "editor is null")
```
Existing code just use error in such a case, you could do the same here for consistency.

MachineThing replied:
I used `Log.wtf` mainly because of the fact that editor shouldn't be null at all. As according to Android documentation:

> `Log.wtf`:
> What a Terrible Failure: Report a condition that should never happen

But I will change it to `Log.e` for consistency

vmiklos commented on app/src/main/java/hu/vmiklos/plees_tracker/Preferences.kt:
```diff
@@ -20,6 +30,76 @@ class Preferences : PreferenceFragmentCompat() {
+                    }
+                }
+            } else {
+                Log.wtf(TAG, "sharedPreferences is null")
```
Same here.

vmiklos commented on app/src/main/java/hu/vmiklos/plees_tracker/Preferences.kt:
```diff
@@ -20,6 +30,76 @@ class Preferences : PreferenceFragmentCompat() {
         }
     }
+
+    override fun onPreferenceTreeClick(preference: Preference): Boolean {
```
We already have other settings that want to do something custom after being enabled, for example when you enable automatic backup. In that case we have a listener in SharedPreferencesChangeListener.kt to perform the custom action. Would you mind moving this logic to there as well, for consistency?

MachineThing replied:
Go for it. I wasn't really sure where to put this code to be honest but yeah for consistency it should be moved there.

vmiklos commented on app/src/main/java/hu/vmiklos/plees_tracker/Preferences.kt:
```diff
@@ -20,6 +30,76 @@ class Preferences : PreferenceFragmentCompat() {
+
+        if (!hasPermission) {
+            if (editor != null) {
+                editor.putBoolean("enable_dnd", false)
```
It would be also an option to just not to do this. The "enabled, but no permission" case is something you have to handle anyway (and you do), so if the user explicitly taps enable + rejects the permission, I would not auto-disable.

vmiklos reviewed (approved):
Thanks, now I read this from start to end, and it does what is advertised and in general the code style is OK, we also have documentation.

I added some inline comments, especially merging your permission logic into SharedPreferencesChangeListener.kt would be nice in a follow-up PR if you are interested. (If not, let me know, I'll try to get to it.)

Last, this has no tests. But I find testing really hard, to it's OK to not do it. In the unlikely case you have some Android testing experience, do look at app/src/androidTest/java/hu/vmiklos/plees_tracker/MainActivityInstrumentedTest.kt, which has a FIXME, which kind of blocks adding more tests.

All in all, nice optional feature. :-)
